pkg/orchestrator/provider/aws/client.go

```diff
@@ -256,7 +256,7 @@ func (c *Client) RunAssetScan(ctx context.Context, config *provider.ScanJobConfi
 
 	// Create scanner instance
 	numOfGoroutines := 2
-	errs := make(chan error, numOfGoroutines)
+	errs := make([]error, numOfGoroutines)
```

Comment:
You shouldn't write into a slice by multiple goroutines.
You can keep it as an unbufferd channel and do something like
```
var errs []error
errorsChan := make(chan error)
go func() {
	for e := range errorsChan {
		errs = append(errs, e)
	}
}()
.
.
.

close(errorsChan)
err = errors.Join(errs...)
```

Reply:
Yes @FrimIdan . Thanks for reminding .

I think we should use buffered channel as we already know how many goroutines we need in advance. 

**Sample:**

```
.
.
numOfGoroutines := 3
var errs []error
errsChan := make(chan error, numOfGoroutines)

var wg sync.WaitGroup

wg.Add(1)
go func() {
defer wg.Done()
.
.
errsChan <- WrapError(fmt.Errorf("failed to delete scanner volume: %w", err))
.
.
}()


wg.Add(1)
go func() {
defer wg.Done()
.
.
errsChan <- WrapError(fmt.Errorf("failed to delete scanner volume snapshot: %w", err))
.
.
}()

wg.Add(1)
go func() {
defer wg.Done()
.
.
errsChan <- fmt.Errorf("failed to delete asset volume snapshot: %w", err).
.
.
}()

wg.Wait()
close(errsChan)
for e := range errsChan {
        if e != nil {
	        // nolint:typecheck
	        errs = append(errs, e)
        }
}
if err != nil {
        errs = append(errs, err)
}
err = errors.Join(errs...)

```

Let me know if above changes are fine or anything else is needed.

Thanks

Reply:
Using a buffered channel assumes that every go routine is going to error and only error once, hopefully none of the go routines will error so allocating a buffered channel allocates memory we don't need.

I think its much cleaner/readable to have a routine draining the error channel into the slice, and then after everything is done close the channels which end the routines, and then you can just join the slice straight away instead of having another loop to run after everything else is completed.

Reply:
**For the first part:**
It seems to me, we can not drain all errors of unbufferred channel at one go.
As per documentaion, If the channel is unbuffered, the sender blocks until the receiver has received the value. 

I have tried to work with sample program below.
It works with buffered channel but not working fine with unbuffered channel (errsChan := make(chan error)). 

**For the second part:** 
yes it's a better approach for a separate gorutine for draining error channel.
I have mentioned below, please check.

```
func main() {

	var errs []error
	errsChan := make(chan error, 4)

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		errsChan <- errors.New("error 1")
		errsChan <- errors.New("error 2")
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()
		errsChan <- errors.New("error 3")
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()
		errsChan <- errors.New("error 4")
	}()
	wg.Wait()
	close(errsChan)

	wg.Add(1)
	go func() {
		defer wg.Done()
		for e := range errsChan {
			errs = append(errs, e)
		}
	}()
	wg.Wait()

	fmt.Print(errs)
}
```

Reply:
Ah I see race condition, basically we need to make sure that the draining routine is done after we close the channel so that there is no race between the `fmt.Print(errs)` and the drain. This code works https://go.dev/play/p/UEUxO-gWIz6 please take a look.

Reply:
Great . Thanks a lot for knowledge sharing . :+1: 

I have pushed changes again.